### PR TITLE
Merge Kotlin DSL source distro transforms for faster first use

### PR DESCRIPTION
**Pull request by eskatos**
* Fixes https://github.com/gradle/gradle/issues/28986
* See details in https://github.com/gradle/gradle/issues/28986#issuecomment-2114283928
* Also see https://github.com/gradle/gradle/issues/30440 for some background on temporary directories in artifact transforms

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Review by h0tk3y — APPROVED** [review]
LGTM overall, see one comment about the temporary directory

File: platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt [hunk 1]
```diff
@@ -22,67 +22,81 @@ import org.gradle.api.artifacts.transform.TransformOutputs
...
+     *
+     * See https://github.com/gradle/gradle/issues/30440
+     */
+    private fun TransformOutputs.withTemporaryDir(name: String, block: (File) -> Unit) {
+        val dir = dir(name)
+        try {
+            block(dir)
+        } finally {
+            dir.deleteRecursively()
+            dir.mkdirs()
```
[the remark this answers is not in the document]
Comment (eskatos) [reply platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt]: Good point, I changed it to ensure the cleanup happens

**Pull request merged**